neuralpp/symbolic/z3_expression.py

```diff
@@ -498,7 +512,115 @@ def __and__(self, other: Any) -> Context:
                 other = Z3Expression.convert(other)
             # Always treat `other` as a literal. Will raise if it cannot be converted to a boolean in Z3.
             new_solver = z3_add_solver_and_literal(self._solver, other.z3_object)
-            new_dict = self._dict | _extract_key_value_from_assertions([other.z3_object])
+            new_dict = self._dict | _extract_key_to_value_from_assertions([other.z3_object])
         return Z3SolverExpression.make(new_solver, new_dict)
 
     __rand__ = __and__
+
+    @cached_property
+    def variable_replacement_dict(self):
```

Comment:
This property is dependent on our new class `EquivalenceClass`, see comments here for more detailed description.